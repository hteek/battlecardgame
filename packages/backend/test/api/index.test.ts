import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Definition } from 'aws-cdk-lib/aws-appsync';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { UserPool } from 'aws-cdk-lib/aws-cognito';

import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

import { ApiStack, BattleCardGameGraphqlApi } from '../../lib/api/index.js';

import { apiName, domainName, hostedZoneId, userPoolId } from '../data/index.js';

describe('Api nested stack', () => {
  const OLD_ENV = process.env;

  beforeEach(() => {
    process.env = { ...OLD_ENV }; // make a copy
  });

  afterEach(() => {
    process.env = OLD_ENV; // restore old env
  });

  test('Test default stack', async () => {
    const stack = new Stack(new App(), 'TestStack');
    const userPool = UserPool.fromUserPoolId(stack, 'TestUserPool', userPoolId);

    const internalApi = new BattleCardGameGraphqlApi(stack, 'TestApiInternal', {
      definition: Definition.fromFile(join(dirname(fileURLToPath(import.meta.url)), 'schema.graphql')),
      name: 'test-api-internal',
      userPool,
    });

    const nestedStack = new ApiStack(stack, 'Api', {
      config: {
        apiName,
        certificate: Certificate.fromCertificateArn(
          stack,
          'Certificate',
          'arn:aws:acm:us-east-1:account:certificate/certificate_id',
        ),
        domainName,
        hostedZoneId,
        sourceApis: [internalApi],
        userPool,
      },
    });
    expect(Template.fromStack(nestedStack).toJSON()).toMatchSnapshot();
  });
});
