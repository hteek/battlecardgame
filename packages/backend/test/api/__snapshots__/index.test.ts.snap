// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Api nested stack > Test default stack 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "referencetoTestStackTestApiInternalB2077A5FApiId": {
      "Type": "String",
    },
  },
  "Resources": {
    "CnameRecord2351B7B9": {
      "Properties": {
        "HostedZoneId": "XXXXXXXXXXXXX",
        "Name": "api.my.custom.domain.",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "DomainName",
              "AppSyncDomainName",
            ],
          },
        ],
        "TTL": "1800",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "DomainName": {
      "Properties": {
        "CertificateArn": "arn:aws:acm:us-east-1:account:certificate/certificate_id",
        "DomainName": "api.my.custom.domain",
      },
      "Type": "AWS::AppSync::DomainName",
    },
    "DomainNameApiAssociation": {
      "DependsOn": [
        "DomainName",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "MergedGraphqlApi55953EF9",
            "ApiId",
          ],
        },
        "DomainName": "api.my.custom.domain",
      },
      "Type": "AWS::AppSync::DomainNameApiAssociation",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MergedGraphqlApi55953EF9": {
      "Properties": {
        "AdditionalAuthenticationProviders": [
          {
            "AuthenticationType": "AWS_IAM",
          },
        ],
        "ApiType": "MERGED",
        "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
        "LogConfig": {
          "CloudWatchLogsRoleArn": {
            "Fn::GetAtt": [
              "MergedGraphqlApiApiLogsRole64F64A93",
              "Arn",
            ],
          },
          "ExcludeVerboseContent": false,
          "FieldLogLevel": "NONE",
        },
        "MergedApiExecutionRoleArn": {
          "Fn::GetAtt": [
            "MergedGraphqlApiMergedApiExecutionRoleA35F71F0",
            "Arn",
          ],
        },
        "Name": "testApi",
        "UserPoolConfig": {
          "AwsRegion": {
            "Ref": "AWS::Region",
          },
          "DefaultAction": "ALLOW",
          "UserPoolId": "eu-central-1_XxXxxxxX1",
        },
        "Visibility": "GLOBAL",
        "XrayEnabled": true,
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "MergedGraphqlApiApiLogsRole64F64A93": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSAppSyncPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MergedGraphqlApiLogRetention7D37CB1E": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/appsync/apis/",
              {
                "Fn::GetAtt": [
                  "MergedGraphqlApi55953EF9",
                  "ApiId",
                ],
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "MergedGraphqlApiMergedApiExecutionRoleA35F71F0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MergedGraphqlApiMergedApiExecutionRoleDefaultPolicy2BCA9D41": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "appsync:SourceGraphQL",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MergedGraphqlApiTestApiInternalAssociationB16E7CB0",
                    "SourceApiArn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MergedGraphqlApiTestApiInternalAssociationB16E7CB0",
                          "SourceApiArn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "appsync:StartSchemaMerge",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MergedGraphqlApiTestApiInternalAssociationB16E7CB0",
                  "AssociationArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MergedGraphqlApiMergedApiExecutionRoleDefaultPolicy2BCA9D41",
        "Roles": [
          {
            "Ref": "MergedGraphqlApiMergedApiExecutionRoleA35F71F0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MergedGraphqlApiTestApiInternalAssociationB16E7CB0": {
      "Properties": {
        "MergedApiIdentifier": {
          "Fn::GetAtt": [
            "MergedGraphqlApi55953EF9",
            "ApiId",
          ],
        },
        "SourceApiAssociationConfig": {
          "MergeType": "AUTO_MERGE",
        },
        "SourceApiIdentifier": {
          "Ref": "referencetoTestStackTestApiInternalB2077A5FApiId",
        },
      },
      "Type": "AWS::AppSync::SourceApiAssociation",
    },
  },
}
`;
